File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> vals = new ArrayList<>();
        getAllVals(vals, root);
        return vals;
    }

    private void getAllVals(List<Integer> vals, TreeNode node){
        if(node == null){
            return;
        }
        vals.add(node.val);
        getAllVals(vals, node.left);
        getAllVals(vals, node.right);
    }
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun preorderTraversal(root: TreeNode?): List<Int> {
        val vals = mutableListOf<Int>()
        getAllVals(vals, root)
        return vals
    }

    private fun getAllVals(vals: MutableList<Int>, node: TreeNode?) {
        if (node == null) return
        vals.add(node.`val`)
        getAllVals(vals, node.left)
        getAllVals(vals, node.right)
    }
}